### PR TITLE
feat(dataset) Split more circular dependencies

### PR DESCRIPTION
**Pull request by fpacifici**
Adding more features to Query structure and query processing makes the Query class more complex (and this is intended). In this scenario the basic util module cannot depend on Query, because that means everyone who tries to import basic functions like to_list and escape_alias would have to depend (indirectly) on Query, which seems a wrong design.
On the other hand, moving a lot of parsing functionalities to Query means Query will have to depends on some utils.
Thus this removes the dependency from util to Query by splitting out the column parsing functions in their own module.

Also we now have an implicit dependency on Dataset every time we need to depend on Schema. This is wrong since Dataset is a massive facade, while lower level components may have to depend on the Schema (like the query itself). This PR also moves the declaration of Dataset in its own module within datasets.

### CODE REVIEW

File: snuba/datasets/dataset.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
+from typing import Optional, Mapping, NamedTuple, Sequence, Tuple, Union
```
Comment (fpacifici) [inline snuba/datasets/dataset.py]: 100% cut and paste from __init__

File: snuba/query/columns.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
+from itertools import chain
```
Comment (fpacifici) [inline snuba/query/columns.py]: I do not expect this module to exist in this form for long. column_expr and similar functions are going to be embedded in column abstractions as soon as they are introduced, but we need to break it off from util to be able to make Query more structured, which is required to introduce Column abstractions, which is required to embed these functions into such abstractions.

As of now 100% cut/paste from util.

**Review by tkaemming — APPROVED** [review]
This change itself looks good, but I'm not sure I understand the relationship between the change and this part of the description:

> Also we now have an implicit dependency on Dataset every time we need to depend on Schema. This is wrong since Dataset is a massive facade, while lower level components may have to depend on the Schema (like the query itself).

I might be misinterpreting it though.

**Comment by fpacifici** [comment]
re: dataset:
schemas are in `/snuba/datasets/schemas`
the Dataset class is declared in `/snuba/datasets/__init__.py`
if I want Query to be able to reference anything in schemas, by importing snuba.datasets.schemas.anything I am also importing Dataset since it is in __init__

And thus circular dependency since Dataset is a very high level concept, while Query is much lower.

**Pull request merged**